anjali/sb-pt-cont/9.js

```diff
@@ -0,0 +1,231 @@
+        noDuplicateCoords = false;
+        break;
+      }
+    }
+    if (!noDuplicateCoords) {
+      break;
+    } else {
+      uniqueStartingSpriteLocations.push(coords);
+    }
+  }
+  if (spriteIds.length == uniqueStartingSpriteLocations.length) {
+    validationProps.successCriteria.differentLocations = true;
+  }
+
+  // Check sprite costumes
+  var uniqueStartingSpriteCostumes = [];
+  for (var i=0; i<spriteIds.length; i++) {
+    var costume = getProp({id: spriteIds[i]}, "costume");
+    var noDuplicateCostumes = true;
+    for (var j=0; j<uniqueStartingSpriteCostumes.length; j++) {
+      if (costume == uniqueStartingSpriteCostumes[j]) {
+        noDuplicateCostumes = false;
+        break;
+      }
+    }
+    if (!noDuplicateCostumes) {
+      break;
+    } else {
+      uniqueStartingSpriteCostumes.push(costume);
+    }
+  }
+  if (spriteIds.length == uniqueStartingSpriteCostumes.length) {
+    validationProps.successCriteria.differentCostumes = true;
+  }
+}
+
+// Check if student is interacting with the program
+// version 1
+for (var spriteId in spriteIds) {
+  if (getProp({id: spriteId}, "timeout")>0) {
+    validationProps.waitTime = World.frameCount + 60;
+    break;
+  } else {
+    validationProps.waitTime = validationProps.previous.waitTime;
+  }
+}
+
+/*
+// version 2
+if (validationProps.successTime && (eventLog.length > validationProps.previous.eventLogLength)) {
+  validationProps.waitTime = World.frameCount + 100;
+} else {
+  validationProps.waitTime = validationProps.previous.waitTime;
+}
+*/
+
+// Keeps track of which sprites were clicked on
+if(!validationProps.spritesClickedOn) {
+  validationProps.spritesClickedOn = [];
+}
+
+// Keeps track of which sprites caused speech on click
+if(!validationProps.spriteClicksCauseSpeech) {
+  validationProps.spriteClicksCauseSpeech = [];
+}
+
+// Helper function
+// returns true if spriteId is in spriteIdArray
+function member(spriteId, spriteIdArray) {
+  for (var i = 0; i < spriteIdArray.length; i++) {
+    if (spriteIdArray[i] == spriteId) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// New event happened
+if (eventLog.length > validationProps.previous.eventLogLength) {
+  // check eventLog for WhenClick or WhileClick event
+  var currentEvent = eventLog[eventLog.length - 1];
+
+  if (currentEvent.includes("whenClick: ") || currentEvent.includes("whileClick: ")) {
+    // current event is a click
+    var clickedSpriteId = parseInt(currentEvent.split(" ")[1]);
+    validationProps.spritesClickedOn.push(clickedSpriteId);
+    validationProps.successCriteria.numSpritesClicked = validationProps.spritesClickedOn.length;
+    if (!member(clickedSpriteId, validationProps.spriteClicksCauseSpeech)) {
+      var newSpriteSpeaking = false;
+      for (var spriteId in spriteIds) {
+        if (getProp({id: spriteId}, "speech") && getProp({id: spriteId}, "timeout")==120) {
+          // new sprite caused speech in some sprite
```

Comment:
(maybe a question for mike), do we care which sprites say something? currently, if you have click events on all the sprites but they all cause the same sprite to say something, you'll pass.
Ex:
![image](https://user-images.githubusercontent.com/8787187/126231481-61141db6-1ff0-483d-b2b3-a5d3c8bec48f.png)

Reply:
Yes, Mike wanted it like this so the student can make one sprite the 'narrator.'